### ISSUE TITLE
Maximum call stack RangeError in 2.4.0

### ISSUE THREAD

**Issue by ajgreenb** [eslint/eslint#5654]
**What version of ESLint are you using?**

2.4.0

**What parser (default, Babel-ESLint, etc.) are you using?**

Default.

**Please show your full configuration:**

```javascript
module.exports = {

  parserOptions: {
    ecmaVersion: 6,
    sourceType: module,
    ecmaFeatures: {
      jsx: true,
    },
  },

  rules: {

    // Allow for non-camel-cased identifiers. Maybe this should be disabled and
    // then only disabled on a case-by-case basis?
    camelcase: 0,

    // Require curly braces for all control statements and blocks.
    curly: 2,

    // Require type-safe equality (=== instead of ==).
    eqeqeq: 2,

    // Enforce consistent indentation (4 spaces by default.)
    indent: 2,

    // Disallow bitwise operators.
    'no-bitwise': 2,

    // Warn for uses of `this` outside classes or objects.
    'no-invalid-this': 1,

    // Disallow weird unicode whitespace characters that can break parsers.
    'no-irregular-whitespace': 2,

    // No end-of-line whitespace.
    'no-trailing-spaces': 2,

    // Disallow variables not declared with var, let, or const.
    'no-undef': 2,

    // Disallow unused variable declarations.
    'no-unused-vars': 2,

    // Require single quotes instead of double quotes (unless they are used
    // inside a singled-quoted string.)
    quotes: [ 2, 'single', 'avoid-escape' ],

    // Require a space before and after JavaScript keywords.
    // I.e., if () {} else {} as opposed to if() {}else{}.
    'keyword-spacing': 2,

    // Require a space before the curly braces that start a block.
    // I.e., function () {} as opposed to function (){}.
    'space-before-blocks': 2,

    // Require a space before the opening parenthesis of an anonymous, but not
    // named, function's parameters.
    // I.e., function (), not function(), and
    // function foo(), not function foo ().
    'space-before-function-paren': [ 2, { anonymous: 'always', named: 'never' } ],

    // Syntactic additions that tell ESLint to lint within JSX code.
    'react/jsx-uses-react': 2,
    'react/jsx-uses-vars': 2,
    'react/react-in-jsx-scope': 2,
  },

  // Defines the environments in which the linted code will run. This specifies
  // certain keywords that ESLint will recognize and allow.
  env: {
    browser: true,
    es6: true,
  },

  plugins: [

    // Make ESLint recognize React/JSX code.
    'react',

  ],

  // Specify that this is the top-level ESLint configuration file, and not to
  // traverse higher.
  root: true,

};
```

**What did you do? Please include the actual source code causing the issue.**

I ran `eslint path\to\Footer.js`.  Here is `Footer.js`:

```javascript
import React from 'react';

/**
 * Components that renders a mini footer. It doesn't do much else
 * except display a title.
 * @component
 * @exports path\to\Footer
 */
const Footer = React.createClass({

    render: function() {
        return (
            <footer className="footer flex-center">
                <h4>A Cool Title</h4>
            </footer>
        );
    }
});

export default Footer;
```

**What did you expect to happen?**

For the code to be linted.

**What actually happened? Please include the actual, raw output from ESLint.**

```
C:\Users\me\project>eslint path\to\Footer.js
Maximum call stack size exceeded
RangeError: Maximum call stack size exceeded
    at C:\path\to\node_modules\eslint\lib\config\config-ops.js:181:48
    at Array.forEach (native)
    at deepmerge (C:\path\to\node_modules\eslint\lib\config\config-ops.js:175:30)
    at C:\path\to\node_modules\eslint\lib\config\config-ops.js:181:32
    at Array.forEach (native)
    at deepmerge (C:\path\to\node_modules\eslint\lib\config\config-ops.js:175:30)
    at C:\path\to\node_modules\eslint\lib\config\config-ops.js:181:32
    at Array.forEach (native)
    at deepmerge (C:\path\to\node_modules\eslint\lib\config\config-ops.js:175:30)
    at C:\path\to\node_modules\eslint\lib\config\config-ops.js:181:32
```

**Comment by ajgreenb** [comment]
Oh, and the version of `eslint-plugin-react` is `^4.2.2`, if that helps.

**Comment by gyandeeps** [comment]
Have you tried it by removing `eslint-plugin-react`?

**Comment by ajgreenb** [comment]
@gyandeeps I just tried after removing `eslint-plugin-react`:

```javascript

...

    // 'react/jsx-uses-react': 2,
    // 'react/jsx-uses-vars': 2,
    // 'react/react-in-jsx-scope': 2,
  },

...

  // plugins: [
  //
  //   // Make ESLint recognize React/JSX code.
  //   'react',
  //
  // ],

...

```

but I still get the same error. I think it has something to do with my rules.

**Comment by gyandeeps** [comment]
this looks odd
```js
sourceType: module,
```

I think you should have quotes around `module`. 
Not sure if this is the real prob or not

**Comment by ilyavolodin** [comment]
The error happens here: https://github.com/eslint/eslint/blob/master/lib/config/config-ops.js#L181
It looks like deepmerge goes into infinite recursion loop. @ajgreenb Could you run eslint with `--debug` flag to see if any other configs are being picked up? The config you posted looks reasonable, maybe an issue is happening in some other config file?

**Comment by ajgreenb** [comment]
It appears that @gyandeeps was correct; the culprit was a missing pair of quotes. Switching to

```javascript
sourceType: 'module',
```

fixed the error. Should I close the issue?

**Comment by ajgreenb** [comment]
Here's the output of ESLint with `--debug`, just to be thorough:

```
C:\path\to\project>eslint --debug components\BriefFooter.js
  eslint:cli Running on files +0ms
  eslint:ignored-paths Looking for ignore file in C:\path\to\project +72ms
  eslint:ignored-paths Could not find ignore file in cwd +3ms
  eslint:cli-engine Processing C:\path\to\project\components\BriefFooter.js +3ms
  eslint:cli-engine Linting C:\path\to\project\components\BriefFooter.js +2ms
  eslint:config Constructing config for C:\path\to\project\components\BriefFooter.js +0ms
  eslint:config Using .eslintrc and package.json files +1ms
  eslint:config Loading C:\path\to\project\.eslintrc.js +4ms
  eslint:config-file Loading JS config file: C:\path\to\project\.eslintrc.js +2ms
Maximum call stack size exceeded
RangeError: Maximum call stack size exceeded
    at C:\path\to\project\node_modules\eslint\lib\config\config-ops.js:181:48
    at Array.forEach (native)
    at deepmerge (C:\path\to\project\node_modules\eslint\lib\config\config-ops.js:175:30)
    at C:\path\to\project\node_modules\eslint\lib\config\config-ops.js:181:32
    at Array.forEach (native)
    at deepmerge (C:\path\to\project\node_modules\eslint\lib\config\config-ops.js:175:30)
    at C:\path\to\project\node_modules\eslint\lib\config\config-ops.js:181:32
    at Array.forEach (native)
    at deepmerge (C:\path\to\project\node_modules\eslint\lib\config\config-ops.js:175:30)
    at C:\path\to\project\node_modules\eslint\lib\config\config-ops.js:181:32
```

**Comment by ilyavolodin** [comment]
Looks like it's a problem with configuration in that case. I'm not sure we can prevent something like that in our code. So closing since it looks like this issue has been resolved.
